operator/docs/enhancements/zone_aware_replication.md

```diff
@@ -0,0 +1,358 @@
+
+This is how the expected individual pod spec will look like after the webhook has added the topology key annotation in the pod:
+
+```yaml
+kind: Pod
+apiVersion: v1
+metadata:
+  name: ingester
+  annotations: 
+    topology.kubernetes.io/zone: zone-a 
+spec:
+  topologySpreadConstraints:
+  - maxSkew: 1
+    topologyKey: topology.kubernetes.io/zone
+    whenUnsatisfiable: DoNotSchedule
+    labelSelector:
+      matchLabels:
+        ingester: pod
+  containers:
+  - name: ingester
+    env:
+    - name: ZONE
+        valueFrom:
+          fieldRef:
+            fieldPath: metadata.annotations['topology.kubernetes.io/zone']
+```
+
+The container can then be started by passing the configuration via the cli flag `instance_availability_zone: $ZONE`
+
+
+### Open Questions
+
+* Both these methods first modify the pod to add the annotation[topology.kubernetes.io/zone: zone-a] in the pod and then modify the container to add a new env var or a volume which picks the zone value from the pod annotation via the downwardapi
+* In the Webhook method, is it safe to assume both the pod modification, followed by the container modification can be handled within the webhook?
+* In the operator method, there is going to be two merge patches of the object, one for the pod to add the annotation, and then the container to add the env variable.
+* Would it be better to handle the multiple modifications in an init container. If yes, Is that worth the overhead of  maintainence of an init container?
+* If the operator can handle this without init container isnt that better than a webhook?
+
+## Implementation History
+
+## Drawbacks
+By just using the Topologykey to deploy the replicas in the different zones, we only ensure that 2 pods of different zones are not in the same node. But we cant control 2 replica pods in the same zone getting deployed on the same node. To prevent this we might want to let the user input zone details that can be used as a NodeSelector
+
+## Alternatives
+ 
+To enable zone-aware replication for the write path and the read path:
+
+* Configure the availability zone for each ingester via the `-ingester.availability-zone` CLI flag (or its respective YAML config option)
+* Rollout ingesters to apply the configured zone
+* Enable time-series zone-aware replication via the `-distributor.zone-awareness-enabled` CLI flag (or its respective YAML config option). Please be aware this configuration option should be set to `distributors`, `queriers` and `rulers`.
+https://github.com/grafana/rollout-operator
+
+<div id="sizing"/>
+
+## Impact on LokiStack Sizing Configurations
+Without zone-aware replication, the LokiStack pods are scheduled on different nodes within the same or different availability zones. There are no hard requirements on the number of zones within the cluster. In case a zone fails, the pods are simply rescheduled to other nodes based on the current policies (taints/tolerations/affinity).
```

Comment:
I don't think the last sentence is true. The scheduler will try to reschedule pods to a different zone, if available, but the issue is with the PersistentVolumes. Usually the PVs created on public clouds are also bound to a zone and can not be attached to a virtual machine (= Kubernetes Node) running in a different zone.

This will effectively mean that the ingestion will stop during a zone outage as the ingester pod can not be rescheduled to a different zone without also moving the PV, which is usually not possible. Kubernetes will not do anything in this case, the Pod will simply stay "Pending". Manual intervention _can_ remove the PVC, in which case it will be recreated by the operator and then possibly lead to a new PV created in a different zone and the Pod starting. But this is a manual operation and will also lead to loss of the data contained in the removed PV.
It might also be impossible to remove the PV while the zone is unavailable.

Reply:
After playing around with their tanka config, their PVC looks like this:

Reply:
I'm not sure what the additional level of indirection changes regarding the situation with the PVs. "Something" needs to step-in and manually remove the PVs, so that they are recreated. In our case this would be a human (current solution) or the loki-operator (possible future solution). In the case of Tanka above this would be the "rollout-operator" then? But what does this operator know about the internal operation of Tanka, so that it knows it is safe to remove the PVs / redistribute to another zone?

Reply:
Mentioned it here - https://github.com/grafana/loki/blob/48ae6628e1508ceac3bfd897cc1ec52c66d651b6/operator/docs/enhancements/zone_aware_replication.md#impact-on-lokistack-sizing-configurations

Reply:
As mentioned in a call, we have two options here:
1. Treat this proposal simply as it says it will provide namely zone-aware data replication only. This means that we need some sort of playbook to execute failover activities during a zone outage.
2. Expand the operator capabilities to active fail-over from one zone to the other.

In summary we have consensus to err on the side of a simpler feature going forward with option no.1. The latter option demands to re-visit the capabilities of the Loki Operator managing StatefulSets (e.g. ingesters) via a custom controller to enable fail-over across zones. This was identified to be a prerequisite to the zone-aware data replication feature as described in the proposal.

Reply:
@cstyan :point_up: thread is interesting for you maybe. Do you use the rollout-operator to do automatic failure due to the PVCs bound in one zone?

Reply:
Recap of our slack conversation for those who might be reading along here:

The short answer is no. Currently we only use "zone awareness" in the sense of logical zone awareness. We use the zone awareness config and separate node pools to replicate each piece of data to one ingester in each zone, which means we can roll out ingester pods much faster during deployments. In some cases you can roll an entire zone at once. The rollout operator, via some labels configuration, knows about the three different statefulsets, and ensures that all pods in a single zone are rolled before moving on to the next one.

To use actual phsyical zone awareness the implementation at the hardware or cloud provider level is likely what would be needed. Some options might be via things like GKE mutli-zone and/or regional clusters, google cloud regional persistent disks, and manually creating a node pool in each zone for each ingester statefulset each of which would have a taint, and then each ingester statefulset can be constrained to that set of nodepools via it's toleration config.